### DISCUSSION TITLE
Rule engine questions

### DISCUSSION THREAD

**Discussion by vampywiz17** [nanomq/nanomq#1761] [Q&A]
Hello there,

possible to create a rule for example i got a payload to topic /aaa it send a different payload to topic /bbb ?

**Comment by JaylinYu** [comment]
Sure, it is just a simple republish, sorry we didn't have a tutorial about this in doc.
@xinyi-xs could u help on this

**Comment by vampywiz17** [comment]
@JaylinYu

Thanks the answer!

First, i try a simple thing, "filter" my payload :) but it not work at now. Here my syntax:

```
rules.repub = {
  rules = [
    {
      server = "mqtt-tcp://192.168.31.105:1883"   # MQTT server address for republishing
      topic = "topic/repub"                  # Topic to republish on
      proto_ver = 4                          # MQTT protocol version
      clientid = "repub_client1"             # Client ID for the republishing client
      keepalive = "60s"                      # Ping interval for the republishing client
      clean_start = true                     # Clean start flag for the republishing client
      username = "username"                  # Username for the republishing client
      password = "passwd"                    # Password for the republishing client
      sql =  "SELECT payload.state as y FROM \"opensprinkler\station\1\" WHERE y != 1"    # SQL clause for the rule
    }
 ]
}    
```

But nothing come to topic/repub (the original payload value are 0 or 1) logically i would like that it only come to repub to value 0

**Reply by JaylinYu** [reply]
Whoops, apologize that nanomq's SQL interpreter is not finished yet. Perhaps "!=" is where it goes wrong.

**Reply by vampywiz17** [reply]
@JaylinYu 

I do some more test.

First i try this:

`SELECT payload.state as shelly1 FROM \"opensprinkler/station/1"`

It working well, it forward the payload a different topic with different payload name (shelly1)

after i try this:

`SELECT payload.state as shelly1 FROM \"opensprinkler/station/1" WHERE shelly1 = 1`

It not working, it say syntax error. Ok i try this:

`SELECT payload.state as shelly1 FROM \"#\" WHERE shelly1 = 1`

now it work, but it listen broker all payload, not only the topic that i want.

and it is only the filter, currently no idea, how to change the payload value. For example, if the `shelly1 = 1 ` then `shelly = 'on'`.

**Comment by vampywiz17** [comment]
i also try with emqx, these syntax work like a charm:

```
SELECT
  CASE WHEN payload.state = 0 THEN 'off'
       WHEN payload.state = 1 THEN 'on'
  ELSE payload.state
  END as shelly1
FROM
  "opensprinkler/station/1"
```

republish:

```
{
  "shelly1": "${shelly1}"
}
```

what (if exist) the nanomq variant of this rule?

**Reply by xinyi-xs** [reply]
The `SELECT CASE` syntax is not supported on NanoMQ. You should replace the query from:
```sql
SELECT payload.state as shelly1 FROM \"opensprinkler/station/1" WHERE shelly1 = 1
```
to:
```sql
SELECT payload.state as shelly1 FROM \"opensprinkler/station/1\" WHERE shelly1 = 1

```
The backslash is missing in the original query.

**Reply by vampywiz17** [reply]
@xinyi-xs 

Thanks! The `THEN` are supported?

**Reply by xinyi-xs** [reply]
The current support is limited to the keywords 'SELECT', 'FROM', 'WHERE', and 'as'. Additional keywords may be considered for support in the future.

**Reply by vampywiz17** [reply]
@xinyi-xs I see. This keywords are possible to change the payload value somehow? for example, in above examle, if shelly1 = 1, the shelly1 new value are `'on'`

**Reply by xinyi-xs** [reply]
No that's unsupported for now.
